File: Python/py-maateen/singly_linked_list.py
class Node:

  def __init__(self, data, next):
    self.data = data
    self.next = next


class SinglyLinkedList:

  def __init__(self):
    self.head = None
    self.length = 0

  def insert(self, data):
    self.head = Node(data, self.head)
    self.length += 1

  def remove_at_index(self, index):
    dummy = Node(None, self.head)
    prev, current = dummy, dummy.next
    check_index = 0
    while current:
      print(index, check_index)
      if index == check_index:
        prev.next = current.next
        current.next = None
      check_index += 1
      prev, current = current, current.next

  def iterate(self):
    node = self.head
    while(node):
      print(node.data)
      node = node.next

  def search(self, data):
    index = 0
    node = self.head
    while node:
      if node.data == data:
        return index
      node = node.next
      index += 1
    return -1

File: Java/HowToCodeDotComBD/src/com/bd/howtocode/java/SinglyLinkedList.java
package com.bd.howtocode.java;

public class SinglyLinkedList<Type> {
    private long size;

    private Node<Type> head;
    private Node<Type> tail;

    public void addFirst(Type value) {
        addFirst(new Node<>(value));
    }

    public void addLast(Type value) {
        addLast(new Node<>(value));
    }

    private void addLast(Node<Type> node) {
        if (size == 0) head = node;
        else tail.setNext(node);

        tail = node;
        size++;
    }

    public void addFirst(Node<Type> node) {
        Node<Type> temp = head;
        head = node;
        head.setNext(temp);

        size++;

        if (size == 1) tail = head;
    }

    public Node<Type> getHead() {
        return head;
    }

    public Node<Type> getTail() {
        return tail;
    }

    public void removeFirst() {
        if (size != 0) {
            head = head.getNext();
            size--;
        }

        if (size == 0) {
            tail = null;
        }
    }

    public void removeLast() {
        if (size != 0) {
            if (size == 1) {
                head = null;
                tail = null;
            } else {
                Node<Type> current = head;

                while (current.getNext() != tail) {
                    current = current.getNext();
                }

                current.setNext(null);
                tail = current;
            }
            size--;
        }
    }

    public Type getFirst() {
        return getHead().getValue();
    }

    public boolean remove(Type type) {
        Node<Type> prev = null;
        Node<Type> current = head;

        while (current != null) {
            if (current.getValue().equals(type)) {
                if (prev != null) {
                    prev.setNext(current.getNext());

                    if (current.getNext() == null) tail = prev;
                    size--;
                } else removeFirst();
                return true;
            }
            prev = current;
            current = current.getNext();
        }
        return false;
    }

    public long getSize() {
        return size;
    }

    public void print() {
        System.out.print("Total elements: " + size + " -> ");
        Node node = head;
        while (node != null) {
            System.out.print(node.getValue().toString() + " , ");
            node = node.getNext();
        }
        System.out.println();
    }

    public void clear() {
        for (Node<Type> x = head; x != null; ) {
            Node<Type> next = x.next;
            x.next = null;
            x.value = null;
            x = next;
        }

        head = tail = null;
        size = 0;
    }
}
